File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/playback/state/specs/GivenAHaltedPlaylistStateManager/AndPlaybackPlaysThroughCompletion/WhenObservingPlayback.java
package com.lasthopesoftware.bluewater.client.playback.state.specs.GivenAHaltedPlaylistStateManager.AndPlaybackPlaysThroughCompletion;

import com.lasthopesoftware.bluewater.client.library.access.ILibraryStorage;
import com.lasthopesoftware.bluewater.client.library.access.ISpecificLibraryProvider;
import com.lasthopesoftware.bluewater.client.library.items.media.files.ServiceFile;
import com.lasthopesoftware.bluewater.client.library.items.media.files.nowplaying.storage.NowPlayingRepository;
import com.lasthopesoftware.bluewater.client.library.repository.Library;
import com.lasthopesoftware.bluewater.client.playback.file.preparation.specs.fakes.FakeDeferredPlaybackPreparerProvider;
import com.lasthopesoftware.bluewater.client.playback.file.volume.IPlaybackHandlerVolumeControllerFactory;
import com.lasthopesoftware.bluewater.client.playback.queues.CompletingFileQueueProvider;
import com.lasthopesoftware.bluewater.client.playback.state.PlaylistManager;
import com.lasthopesoftware.bluewater.client.playback.state.bootstrap.PlaylistPlaybackBootstrapper;
import com.lasthopesoftware.bluewater.client.playback.state.volume.PlaylistVolumeManager;
import com.lasthopesoftware.promises.Promise;

import org.junit.BeforeClass;
import org.junit.Test;

import java.io.IOException;
import java.util.Arrays;
import java.util.Collections;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

/**
 * Created by david on 3/11/17.
 */

public class WhenObservingPlayback {

	private static boolean isPlaying;

	@BeforeClass
	public static void context() throws IOException, InterruptedException {
		final FakeDeferredPlaybackPreparerProvider fakePlaybackPreparerProvider = new FakeDeferredPlaybackPreparerProvider();

		final Library library = new Library();
		library.setId(1);

		final ISpecificLibraryProvider libraryProvider = mock(ISpecificLibraryProvider.class);
		when(libraryProvider.getLibrary()).thenReturn(new Promise<>(library));

		final ILibraryStorage libraryStorage = mock(ILibraryStorage.class);
		when(libraryStorage.saveLibrary(any())).thenReturn(new Promise<>(library));

		final PlaylistManager playlistManager = new PlaylistManager(
			fakePlaybackPreparerProvider,
			Collections.singletonList(new CompletingFileQueueProvider()),
			new NowPlayingRepository(libraryProvider, libraryStorage),
			new PlaylistPlaybackBootstrapper(new PlaylistVolumeManager(1.0f), mock(IPlaybackHandlerVolumeControllerFactory.class)));

		final CountDownLatch countDownLatch = new CountDownLatch(5);

		playlistManager
			.startPlaylist(
				Arrays.asList(
					new ServiceFile(1),
					new ServiceFile(2),
					new ServiceFile(3),
					new ServiceFile(4),
					new ServiceFile(5)), 0, 0)
			.next(obs ->  obs.subscribe(p -> {
				countDownLatch.countDown();
			}));

		fakePlaybackPreparerProvider.deferredResolution.resolve().resolve();
		fakePlaybackPreparerProvider.deferredResolution.resolve().resolve();
		fakePlaybackPreparerProvider.deferredResolution.resolve().resolve();
		fakePlaybackPreparerProvider.deferredResolution.resolve().resolve();
		fakePlaybackPreparerProvider.deferredResolution.resolve().resolve();

		countDownLatch.await(1, TimeUnit.SECONDS);

		isPlaying = playlistManager.isPlaying();
	}

	@Test
	public void thenThePlaylistIsNotPlaying() {
		assertThat(isPlaying).isFalse();
	}
}

File: projectBlueWater/src/test/java/com/lasthopesoftware/bluewater/client/playback/engine/GivenAHaltedPlaybackEngine/AndPlaybackPlaysThroughCompletion/WhenObservingPlayback.kt
package com.lasthopesoftware.bluewater.client.playback.engine.GivenAHaltedPlaybackEngine.AndPlaybackPlaysThroughCompletion

import com.lasthopesoftware.bluewater.client.browsing.files.ServiceFile
import com.lasthopesoftware.bluewater.client.browsing.library.access.FakeLibraryRepository
import com.lasthopesoftware.bluewater.client.browsing.library.repository.Library
import com.lasthopesoftware.bluewater.client.browsing.library.repository.LibraryId
import com.lasthopesoftware.bluewater.client.connection.selected.GivenANullConnection.AndTheSelectedLibraryChanges.FakeSelectedLibraryProvider
import com.lasthopesoftware.bluewater.client.playback.engine.PlaybackEngine
import com.lasthopesoftware.bluewater.client.playback.engine.bootstrap.ManagedPlaylistPlayer
import com.lasthopesoftware.bluewater.client.playback.engine.preparation.PreparedPlaybackQueueResourceManagement
import com.lasthopesoftware.bluewater.client.playback.file.PositionedPlayingFile
import com.lasthopesoftware.bluewater.client.playback.file.preparation.FakeDeferredPlayableFilePreparationSourceProvider
import com.lasthopesoftware.bluewater.client.playback.file.preparation.queues.CompletingFileQueueProvider
import com.lasthopesoftware.bluewater.client.playback.nowplaying.storage.NowPlayingRepository
import com.lasthopesoftware.bluewater.client.playback.volume.PlaylistVolumeManager
import com.lasthopesoftware.bluewater.shared.promises.extensions.toExpiringFuture
import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit

private const val libraryId = 455

class WhenObservingPlayback {
	private val mut by lazy {
		val fakePlaybackPreparerProvider = FakeDeferredPlayableFilePreparationSourceProvider()
		val library = Library(id = libraryId)
		val libraryProvider = FakeLibraryRepository(library)
		val preparedPlaybackQueueResourceManagement = PreparedPlaybackQueueResourceManagement(
			fakePlaybackPreparerProvider,
			mockk {
				every { maxQueueSize } returns 1
			}
		)
		val nowPlayingRepository = NowPlayingRepository(
			FakeSelectedLibraryProvider(),
			libraryProvider,
		)
		val playbackBootstrapper = ManagedPlaylistPlayer(
			PlaylistVolumeManager(1.0f),
			preparedPlaybackQueueResourceManagement,
			nowPlayingRepository,
			listOf(CompletingFileQueueProvider()),
		)
		val playbackEngine =
			PlaybackEngine(
				preparedPlaybackQueueResourceManagement,
				listOf(CompletingFileQueueProvider()),
				nowPlayingRepository,
				playbackBootstrapper,
				playbackBootstrapper,
			)

		Pair(fakePlaybackPreparerProvider, playbackEngine)
	}

	private var isPlaying = false
	private var firstPlayingFile: PositionedPlayingFile? = null
	private var isCompleted = false
	private var playbackStarted = false

	@BeforeAll
	fun act() {
		val (fakePlaybackPreparerProvider, playbackEngine) = mut

		val countDownLatch = CountDownLatch(6)
		playbackEngine
			.setOnPlaybackStarted { playbackStarted = true }
			.setOnPlayingFileChanged { _, p ->
				if (firstPlayingFile == null) firstPlayingFile = p
				countDownLatch.countDown()
			}
			.setOnPlaybackCompleted {
				isCompleted = true
				countDownLatch.countDown()
			}
			.startPlaylist(
				LibraryId(libraryId),
				listOf(
					ServiceFile("1"),
					ServiceFile("2"),
					ServiceFile("3"),
					ServiceFile("4"),
					ServiceFile("5")
				),
				0
			)
			.toExpiringFuture()
			.get()
		var playingPlaybackHandler = fakePlaybackPreparerProvider.deferredResolution.resolve()
		for (i in 0..3) {
			val newPlayingPlaybackHandler =
				fakePlaybackPreparerProvider.deferredResolution.resolve()
			playingPlaybackHandler.resolve()
			playingPlaybackHandler = newPlayingPlaybackHandler
		}
		playingPlaybackHandler.resolve()
		countDownLatch.await(1, TimeUnit.SECONDS)
		isPlaying = playbackEngine.isPlaying
	}

	@Test
	fun `then playback is started`() {
		assertThat(playbackStarted).isTrue
	}

	@Test
	fun `then the first playing file is the first service file`() {
		assertThat(firstPlayingFile!!.serviceFile).isEqualTo(ServiceFile("1"))
	}

	@Test
	fun `then the playlist is not playing`() {
		assertThat(isPlaying).isFalse
	}

	@Test
	fun `then the playback is completed`() {
		assertThat(isCompleted).isTrue
	}
}
